src/mono/mono/mini/method-to-ir.c

```diff
@@ -7565,12 +7567,25 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 							}
 						} else {
 							// For structs that cannot be lowered, we change the argument to byref type
-							ptype = mono_class_get_byref_type (mono_defaults.typed_reference_class);
-							// Load the address of the struct
+							if (gklass && (gklass->container_class == swift_self_t))
+								ptype = mono_class_get_byref_type (swift_self);
+							else
+								ptype = mono_class_get_byref_type (mono_defaults.int_class);
+
 							MonoInst *struct_base_address =  mono_compile_create_var (cfg, mono_get_int_type (), OP_LOCAL);
 							CHECK_ARG (idx_param);
 							NEW_ARGLOADA (cfg, struct_base_address, idx_param);
 							MONO_ADD_INS (cfg->cbb, struct_base_address);
+
+							// Load the address of the struct
+							if (gklass && (gklass->container_class == swift_self_t))
+							{
+								MonoClassField *klass_fields = m_class_get_fields (klass);
+								MonoInst *swift_self_base_address = struct_base_address;
+								struct_base_address = mono_compile_create_var (cfg, mono_get_int_type (), OP_LOCAL);
+								MONO_EMIT_NEW_LOAD_MEMBASE_OP (cfg, OP_ADD_IMM, struct_base_address->dreg, swift_self_base_address->dreg, klass_fields->offset);
```

Comment:
Isn't ` klass_fields->offset` always 0 ?

Reply:
Yes, you are right. I was missing `sizeof (MonoObject)`.

Reply:
What's the difference between `sizeof (MonoObject)` and `MONO_ABI_SIZEOF (MonoObject)`? I've been seeing more of the latter in when dealing with loading structs so I wonder if it's the same here...

Reply:
I think the only difference is that `MONO_ABI_SIZEOF (MonoObject)` works with cross compiler. So if you are say on arm64 and you aot compile code for an arm target, `sizeof (MonoObject)` is 16 but in generated code you need 8 since you emit code for arm target. In this context it should be `MONO_ABI_SIZEOF (MonoObject)`.

Loading a field from a vt type is offset with `sizeof (MonoObject)` only if the vt is boxed. This looks fishy to me here. Are you sure this code path is being hit and tested properly ?

Reply:
So `SwiftSelf<T>` is a struct containing a single field of type `T`. Therefore the `T` field is stored at offset 0 in the struct. This struct is passed around by value in C# code, without any boxing happening. Boxing means that the vt is converted to an object and its value copied to that object. Unlike normal valuetypes, objects on the heap contain a sync block, which is this `_MonoObject` header. If you want to access the value of the `T` field in a boxed instance, then you would need to offset by the size of this header. I don't see swift interoping dealing with any boxing.

If I'm correct and this code is wrong then this also raises the question on why no tests are failing. Maybe this path is not properly tested and it needs further attention ?

Reply:
You are right. The `MONO_EMIT_NEW_LOAD_MEMBASE_OP` was used with non-load instructions. I verified this change locally and confirmed that this code path is being executed. It works with an offset of 0.